src/governance-validators.cpp

```diff
@@ -0,0 +1,349 @@
+// Copyright (c) 2014-2017 The Dash Core developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "governance-validators.h"
+
+#include "base58.h"
+#include "utilstrencodings.h"
+
+#include <algorithm>
+
+CProposalValidator::CProposalValidator(const std::string& strDataHexIn)
+    : strData(),
+      objJSON(UniValue::VOBJ),
+      fJSONValid(false),
+      strErrorMessages()
+{
+    if(!strDataHexIn.empty()) {
+        SetHexData(strDataHexIn);
+    }
+}
+
+void CProposalValidator::Clear()
+{
+    strData = std::string();
+    objJSON = UniValue(UniValue::VOBJ);
+    fJSONValid = false;
+    strErrorMessages = std::string();
+}
+
+void CProposalValidator::SetHexData(const std::string& strDataHexIn)
+{
+    std::vector<unsigned char> v = ParseHex(strDataHexIn);
+    strData = std::string(v.begin(), v.end());
+    ParseJSONData();
+}
+
+bool CProposalValidator::Validate()
+{
+    if(!ValidateJSON()) {
+        return false;
+    }
+    if(!ValidateName()) {
+        strErrorMessages += "Invalid name;";
+        return false;
+    }
+    if(!ValidateStartEndEpoch()) {
+        strErrorMessages += "Invalid start:end range;";
+        return false;
+    }
+    if(!ValidatePaymentAmount()) {
+        strErrorMessages += "Invalid payment amount;";
+        return false;
+    }
+    if(!ValidatePaymentAddress()) {
+        strErrorMessages += "Invalid payment address;";
+        return false;
+    }
+    if(!ValidateURL()) {
+        strErrorMessages += "Invalid URL;";
+        return false;
+    }
+    return true;
+}
+
+bool CProposalValidator::ValidateJSON()
+{
+    return fJSONValid;
+}
+
+bool CProposalValidator::ValidateName()
+{
+    std::string strName;
+    if(!GetDataValue("name", strName)) {
+        strErrorMessages += "name field not found;";
+        return false;
+    }
+
+    std::string strNameStripped = StripWhitespace(strName);
+
+    if(strNameStripped.empty()) {
+        strErrorMessages += "name is empty;";
+        return false;
+    }
+
+    static const std::string strAllowedChars = "-_abcdefghijklmnopqrstuvwxyz012345789";
```

Comment:
I disagree, the safety advantages of standardizing on std::string far outweigh the negligible cost of the small extra memory allocation.